[Woody] 미션 4 : 모든 기물 배치하기

jihye-woo opened the pull request:
안녕하세요 Woody 입니다! 미션 4 PR 드립니다!

---

### 🤔 아직 고민 중인 부분

#### <del>1.  길게 열거된 코드 리팩토링 </del> ✔ 미션 5 요구사항으로 해결!
-  Piece 클래스의 팩토리 메소드
   - 색상과 Piece 종류에 따라 메소드를 하나씩 구현하다 보니 총 12개의 팩토리 메소드를 구현하게 되었습니다 (🤦‍♀️)
   - 모든 종류의 Piece 객체에 대응하여 메소드를 구현하다보니 코드가 길게 늘어지게 되었습니다.

-  Board 클래스의 initPieces() 메소드
   -  마찬가지로 Piece 클래스의 팩토리 메소드를 호출하는 코드가 들어있다 보니 코드가 함께 길어지게 되었습니다 (🤦‍♀️222)
   - Pawn 이외의 체스말들은 initPieces()에서 한 줄씩 초기화를 해주기 때문에 메소드의 크기가 너무 커져버렸습니다.

<del>- 아직 어떤 식으로 코드를 간결하게 만들어줄지 고민 중입니다..! 혹시 힌트를 주실 수 있으시다면! 감사히 받겠습니다..😅🙇‍♀️</del>

 ✔ 미션 5 요구사항에 위 고민과 관련된 힌트가 이미 나와있네요!! 미션 5에서 팩토리 메소드 리펙토링을 진행하려고 합니다!

#### 2. 이전 단계에서는 사용했지만 지금은 중요하지 않은 메소드, 삭제해야 할까요?
   - Board 클래스의 getWhitePawnsResult() 와 getBlackPawnsResult()
      - 미션 3에서 Pawn만 초기화한 이후 Pawn 라인 상태를 확인하는 용도로 만들어진 메소드입니다.
      - 현재는 print() 메소드로 보드 전체 상태를 확인하는 과정에서 Pawn 라인까지 확인할 수 있기 때문에 활용도가 낮아졌습니다.
   - 삭제해도 될까요?

---

### 🤹‍♀️ 구현 및 수정 내용
⁕ 요구사항에 있는 내용은 제외하고 특별히 제가 더 구현하거나 수정한 내용만 적어놓았습니다.
⁕ 다 읽어보지 않으셔도 괜찮습니다! 코드 이해에 도움이 될까싶어서 적어보았습니다!

#### 1. Representation enum 클래스
   - 체스말의 종류가 다양해지면서, 각 체스말에 대한 representation 정보를 Representation enum 클래스에 따로 분리

#### 2. Board 클래스
   - initialize() 보드 초기화 단계 
     - 색상에 따라 로직 분리
   - add() 매개변수 추가
      -  rank 파라미터 추가 

#### 3. Piece 클래스
   - 팩토리 메소드 추가
      -  Pawn 뿐만 아니라 다른 체스 말에 대한 팩토리 메소드도 추가
      - 색상별, 체스 종류별로 달리하여 총 12개의 메소드 구현
  - 초기화 대상 보드 행(Rank 0 번째, 7번째) 에 대한 클래스 변수 추가
      -  BLACK_PIECES_RANK = 0, WHITE_PIECES_RANK = 7

#### 4. BoardTest 클래스
   - Board 테스트 클래스 추가 
      -  checkTheNumberOfPieces() - 보드 초기화 이후 총 체스말의 갯수를 테스트하는 메소드
   -  테스트 메소드 실행 전 보드 초기화 
      - BeforeEach에 initalize() 호출 코드 추가

#### 5. PieceTest 클래스
   - Piece 기본 생성자가 사라지고 팩토리 메소드로 Piece 객체를 생성하게 되면서 일부 테스트 메소드를 삭제
      - createWithDefaultPawn() 삭제
      - checkThePieceColor() 삭제

---

읽어주셔서 감사합니다!! 🙇‍♀️🙇‍♀️
코드리뷰 해주셔서 (미리) 감사드립니다!!😁😁

jihye-woo committed:
refactor : 람다식 for-loop 수정

  > 람다식의 장점을 활용하지 못하는 부분은 for-loop으로 수정
    - Board 클래스 생성자
    - Board 클래스 initialize() 메소드

jihye-woo committed:
refactor : CommandUtil 클래스 메소드명 수정

  > readCommand() 메소드명 수정
    - 메소드명을 isGameOver()로 하는 것이 흐름상 더 적절함

jihye-woo committed:
docs : 미션 4 체크리스트 추가

  > 미션4.md 추가

jihye-woo committed:
edit : new line 처리 코드 변경

  > 운영체제에 독립적인 new line 처리
    - '\n' → System.getProperty("line.separator") 변경

  > BoardTest의 print() 테스트 확인
    - 테스트 실행 후 test passed 확인

jihye-woo committed:
feat : StringUtils 클래스 구현

  > appendNewLine() 클래스 메소드
    - 인자로 받은 문자열에 줄바꿈 문자열을 추가
    - 새로 만들어진 문자열 리턴

  > NEWLINE 클래스 변수
    - 운영체제에 독립적인 줄바꿈 문자열

  > 싱글턴
    - private 생성자

jihye-woo committed:
refactor : appendNewLine() 적용

  > 줄바꿈 처리 부분 코드 변경
    - StringUtils 클래스 메소드 appendNewLine() 적용

  > BoardTest 클래스 테스트 메소드 print() 확인
    - 테스트 실행 후 test passed 확인

jihye-woo committed:
refactor : Pawn → Piece 클래스명 변경

  > Pawn 클래스 명 변경
    - Piece 클래스

  > 사용하지 않는 메소드 제거
    - Piece 클래스의 getColorName() 메소드 삭제

jihye-woo committed:
edit : Piece 클래스 생성자 수정

  > Piece 클래스 생성자
    - 기본 생성자 삭제
    - 생성자 접근지시자 private 변경

  > PieceTest 테스트 클래스
    - 기본 생성자 관련 테스트 메소드 삭제
    - Piece 생성자가 private 이기 때문에 컴파일 에러 발생

jihye-woo committed:
edit : PieceTest 테스트 클래스 메소드 수정

  > 테스트 메소드 삭제
   - Piece 생성자 수정으로 인해 사용 불가능한 테스트 메소드 삭제

  > 테스트 메소드 추가 및 변경
    - createWhiteAndBlackPieces()
      - Piece 클래스의 팩토리 메소드 구현을 위한 테스트 메소드 작성
    - verifyPawn() 메소드 매개변수 변경

  > 테스트 실행 후 tests failed 확인

jihye-woo committed:
edit : PieceTest 테스트 클래스 메소드 수정

  > 테스트 메소드 삭제
   - Piece 생성자 수정으로 인해 사용 불가능한 테스트 메소드 삭제

  > 테스트 메소드 추가 및 변경
    - createWhiteAndBlackPieces()
      - Piece 클래스의 팩토리 메소드 구현을 위한 테스트 메소드 작성
    - verifyPawn() 메소드 매개변수 변경

  > 테스트 실행 후 tests failed 확인

jihye-woo committed:
edit : Piece 클래스 팩토리 메소드 구현

  > createBlackPawn(), createWhitePawn() 구현
     - 외부에서 new 키워드를 사용하지 않고 Piece 객체 생성 가능
     - new Piece() 사용 코드 → 팩토리 메소드 호출 코드로 변경

  > Piece 클래스 생성자 매개변수 변경
    - Color 타입 color
    - char 타입 representation

  > 테스트 실행 후 tests passed 확인

jihye-woo committed:
feat : Representation enum 클래스 구현

  > Representation enum
    - representation

  > value()
    - Representation 열거 상수가 가진 representation 값을 리턴

jihye-woo committed:
refactor : 사용하지 않는 코드 제거 및 TODO 작성

  > 리펙토링 과정에 수정할 부분 TODO 표시

  > 사용하지 않는 import 제거

  > BoardTest 테스트 클래스 수정
    - print() 테스트 메소드 출력 부분 제거

jihye-woo committed:
edit : BoardTest 테스트 클래스 print() 메소드 수정

  > 예상되는 보드 상태 변경
    - board의 처음과 마지막 Rank까지 초기화 된 상태로 수정

  > 테스트 실행 후 tests failed 확인

jihye-woo committed:
feat : Representation enum 클래스 메소드 추가

  > Representation enum 클래스 메소드 추가
    - white(), black() 메소드 추가

  > Color enum, representation 필드 제거
    - 이후 representation 필드 제거

  > Piece 클래스 representation 관련 코드에 적용
    - white()와 black() 적용

jihye-woo committed:
feat : Piece 클래스, 나머지 체스말에 대한 팩토리 메소드 구현

  > Pawn 이외의 체스말에 대한 팩토리 메소드 구현
    - King, Queen, Knight, Rook, Bishop 객체를 리턴하는 팩토리 메소드 구현

jihye-woo committed:
edit : Board 클래스 initialize() 메소드 수정

  > Board 클래스 initPieces() 구현
    - 각 색상별 첫 번째 Rank를 초기화하는 메소드
    - Piece 클래스에 rank 관련 클래스 변수 추가
      → BLACK_PIECES_RANK, WHITE_PIECES_RANK

  > Board 클래스 initPawns() 구현
    - 각 색상별 Pawn Rank를 초기화하는 메소드

  > initialize()에서 initPawns()와 initPieces() 호출

  > 테스트 실행 후 tests passed 확인

jihye-woo committed:
feat : PieceTest 클래스, 체스말 색상 확인 테스트 메소드 추가

  > checkThePieceColor() 테스트 메소드 구현
    - Piece 클래스의 isWhite(), isBlack() 테스트
    - 생성한 Piece의 색상을 확인

  > 테스트 실행 후 tests failed 확인

jihye-woo committed:
feat : Piece 클래스의 isWhite(), isBlack() 구현

  > isWhite()
    - 현재 Piece의 색상이 흰색이면 true

  > isBlack()
    - 현재 Piece의 색상이 검은색이면 true

  > 테스트 실행 후 tests passed 확인

jihye-woo committed:
refactor: CommandUtil 클래스명, 생성자 접근 지시자 수정

  > util 클래스 이름의 통일성을 위해 클래스명 변경
    - CommandUtil → CommandUtils

  > private 생성자로 변경

jihye-woo committed:
edit : BoardTest 클래스 메소드 수정

  > 테스트 메소드 실행 전
    - 초기화 과정 추가

  > 초기화 후 체스말 수를 체크하는 메소드 추가
    - checkTheNumberOfPieces() 테스트 메소드 추가

  > addThePawnProperly() 시그니처 변경
    - addThePieceProperly()으로 메소드명 수정
    - 매개변수 file 추가(체스의 file 정보)

  > 테스트 메소드명 변경
   - initialize() → initializeBoard()
   - print() → printBoard()

jihye-woo committed:
edit : Color enum 필드 수정

  > Color enum 클래스 필드 추가
   - representation 함수 다시 추가

  > Representation enum 메소드 제거
   - white(), black() 제거

jihye-woo committed:
feat : PieceTest 클래스, 팩토리 메소드 테스트 추가

  > createWhiteAndBlackPieces() 색상별로 분리
    - createBlackPieces() 테스트 메소드 추가
      → 검은색 체스말 생성 팩토리 메소드 테스트
      → 헬퍼 메소드 verifyBlackPiece() 구현
    - createWhitePieces() 테스트 메소드 추가
      → 흰색 체스말 생성 팩토리 메소드 테스트
      → 헬퍼 메소드 verifyWhitePiece() 구현

jihye-woo committed:
refactor : StringUtils 클래스 변수 접근 지시자 변경

  > 클래스 변수 NEWLINE 접근지시자 private 변경

jihye-woo committed:
refactor : Game 클래스 try-with-resources 적용

  > Scanner 사용 구간, try-with-resources 적용
    - 명시적으로 close() 하는 코드 삭제

kyupid commented on src/main/java/net/woody/chess/Board.java:
```diff
@@ -15,47 +15,74 @@
-        return (newPawn.getColor() == Color.WHITE) ? Pawn.WHITE_PAWN_RANK : Pawn.BLACK_PAWN_RANK;
-    }
-
     public Rank getRank(int rank) {
         if (rank < 0 || board.size() <= rank) {
             throw new ArrayIndexOutOfBoundsException("Rank number " + rank + " is out of range!");
         }
```
_Rank_ 혹시 제가 생각하는 그 랭킹인가요?? ㅎㅎ 일단 계속 스크롤 해보겠습니다

kyupid replied:
쿠퍼도 예외처리하시던데 우디도 하셨네요..전 생각도 못했는데 😂

jihye-woo replied:
아앗 안그래도 rank랑 file이라는 단어가 익숙하지 않은 분들이 많아서 Row랑 Column 같이 직관적으로 이해할 수 있는 네이밍을 하는 게 더 좋을 것 같더라구요 이번에 반영을 못했네요! Kyu 덕분에 다시 리마인드 했습니다 감사합니다😁😁

jihye-woo replied:
예외처리! 저에게도 아직 힘든 부분이에요ㅠㅠ 
어디에 어떤식으로 예외처리할지 보통 생각을 잘 못하는데 이번에는 운이 좋았네요..!👍 
다음에는 운이 아니라 실력이 될 수 있도록 계속 의식적으로 예외처리를 생각해야겠어요 저도 Kyu도 모두모두 파이팅👊😂

kyupid commented on src/main/java/net/woody/chess/Board.java:
```diff
@@ -15,47 +15,74 @@
+        whitePieces.add(Piece.createWhiteBishop());
+        whitePieces.add(Piece.createWhiteKnight());
+        whitePieces.add(Piece.createWhiteRook());
+        size += 16;
```
getSize 메서드를 만드신게 아니라 여기서 아예 size를 정해버리는군요

jihye-woo replied:
넵넵 현재는 size 필드 변수가 존재하고 Piece를 보드에 추가할 때마다 size를 직접 업데이트 시켜줍니다ㅠ Kyu가 저의 Rank 코드를 보셨다고 해서 말씀드리는 거지만..!
- getSize()같은 메소드가 콜이 될 때마다 `List<Rank>`에서 모든 Rank의 Piece 리스트 size를 가져와서 일일히 더하는 방법도 생각해봤는데, 그건 getSize()가 호출될 때마다 8번씩(Rank 수만큼) 순회해야하니까 적용하기 망설여지더라구요
- size와 관련해서 다시 생각해보니 Board 클래스에 있는 add() 메소드를 잘 활용하면 size() 업데이트의 책임을 add() 메소드에게만 줄 수 있을 수도 있겠네요! 이 부분은 좀 더 고민해봐야겠습니다 좋은 포인트 감사드립니다!🙇‍♀️

kyupid commented on src/main/java/net/woody/chess/Board.java:
```diff
@@ -15,47 +15,74 @@
+    }
+
+    public void initPawns() {
+        Rank blackPawns = getRank(Piece.BLACK_PAWN_RANK);
+        Rank whitePawns = getRank(Piece.WHITE_PAWN_RANK);
```
`Rank` 일단 Rank라는 클래스 먼저 살펴보러 가야겠어요

jihye-woo replied:
네이밍 수정이 필요하겠군요 역시ㅠㅠ

kyupid commented on src/main/java/net/woody/pieces/Piece.java:
```diff
@@ -0,0 +1,82 @@
+        return representation;
+    }
+
+    public boolean isWhite() {
+        return color == Color.WHITE;
+    }
+
+    public boolean isBlack() {
+        return color == Color.BLACK;
```
여기 equals로 비교해줘야하는거아닌가용?

tmdgusya replied:
Enum 같은 경우는 싱글톤이라 == 으로도 비교가 가능합니다! 😁

jihye-woo replied:
넵넵 로치말대로 Enum 객체는 하나밖에 없다는게 보장되니까 == 비교를 해주었습니다!

ksundong replied:
이 부분은 조금 고려해봐야 할 것 같아요.
말씀하신 내용이 틀리진 않았는데, '싱글톤이라 == 비교가 가능하다.' 이 말이 조금 애매하게 읽히네요.
저도 좀 더 정리를 해봐야 할 것 같습니다~

ksundong replied:
자칫 잘못하면 '메모리 주소가 같은 것이 싱글턴이다.' 라고 보일 여지가 있는 것 같습니다.

tmdgusya replied:
@ksundong 이런 부분은 생각해본적이 없었는데 확실히 그렇게 해석될수도 있겠군요.
조금 더 좋은 표현을 찾아봐야겠어요 !

kyupid commented on src/main/java/net/woody/pieces/Piece.java:
```diff
@@ -0,0 +1,82 @@
+        return color == Color.BLACK;
+    }
+
+    public static Piece createBlackPawn() {
+        return new Piece(Color.BLACK, Representation.PAWN.value());
+    }
+
+    public static Piece createWhitePawn() {
+        return new Piece(Color.WHITE, Representation.PAWN.value());
```
Color 클래스에서 toLower, toUpper 로 Representation 내에 상수가 가지고있는 representation을 바꿔주는거같은데 코드는 어떻게 읽어야할지 잘몰라서 이부분이신기하네요
그냥 컬러에따라서 white black 관계없이 똑같이 value() 하면 바로 바뀌어서 나오는모양이군요

jihye-woo replied:
앗 아뇨아뇨 value()에서는 representation 소문자 형태로 똑같은 값이 나옵니다! 
대신 Piece 생성자 내부에서 Color enum 클래스의 메소드를 통해 color에 따라 representation을 변경해요! (아래 화살표 지점!!)

```java
   private Piece(Color color, char representation) {
        this.color = color;
        this.representation = color.represent(representation); // 👈👈
    }
```

jihye-woo replied:
[ TMI ] 사실 이 구현부도 많이 고민을 했던 부분입니다..! 
처음에는 Color에서는 representation과 관련된 상태(represent 함수)를 제거하고, Representation 클래스에 아래와 같이 색상별로 다른 representation을 리턴할 수 있도록 코드를 작성하기도 했었거든요!
```java
  // Representation enum 클래스
   public char white() {
        return representation;
    }

    public char black() {
        return Character.toUpperCase(representation);
    }
```
그러면 코드를 아래와 같이 만들 수 있더라구요! 
```java
  // Piece 클래스
    public static Piece createBlackPawn() {
        return new Piece(Color.BLACK, Representation.PAWN.black());
    }
```
블랙블랙 조금 더 직관적이기도 하고 Color의 함수 매개변수를 없앨 수 있어서 가독성이 좋아지더라구용..! 어떤 방식을 선택할지 고민을 좀 했습니다 😂 어떤 전략을 선택해야하는지가 미션을 통틀어 가장 어려운 부분인 것 같아요 엉엉

kyupid replied:
와 완전 똑똑하신거같아요 ㅋㅋㅋ 저렇게해서 소문자나 대문자로 판별하는거군용👏👏👏

jihye-woo replied:
저는 오히려 Kyu처럼 색상별로 enum을 다 만드는 방식을 생각 못했는데 그게 더 깔끔하고 똑똑한 방법이라는 생각이 드는걸요...?👀 👍👍

jihye-woo replied:
근데 미션 5 진행해보니까, enum 관련한 요구사항을 따라가려면 약간의 수정이 필요할 것 같습니다ㅠㅠ

kyupid commented on src/main/java/net/woody/utils/CommandUtils.java:
```diff
@@ -10,7 +13,7 @@ public static void printMenu() {
         System.out.print("> ");
```
첨에 시작할때 대소문자까지 구분해서 요런것까지 신경쓰셨군요 👏

jihye-woo replied:
감사합니당😀 
저같이 어디로 튈지 모르는 사용자가 인풋을 작성하게 된다면 분명 대소문자 오류 한 번 이상은 나지 않을까 싶었어요ㅋㅋㅋㅋ

kyupid reviewed (commented):
안녕하세요 우디~~~한번팀하고 한번도 못만났네요.
진짜 엄청 열심히 코드작성하신 느낌입니다..
저는 그냥 가볍게 한마디씩 감상평 남겨봤습니다!

jihye-woo committed:
refactor : Board 클래스 initialize() 단순화

  > initPieces() 로직을 색상에 따라 분리
   - initBlackPieces()와 initWhitePieces()로 분리

pbg0205 commented on src/main/java/net/woody/utils/StringUtils.java:
```diff
@@ -0,0 +1,17 @@
+    private StringUtils() {
+    }
+
+    public static String appendNewLine(String line) {
```
저는 String으로 사용했는데 StringBuilder 인스턴스 생성하는 방법이 더 효율적인거 같아요👍👍

ksundong commented on checklist/미션4.md:
````diff
@@ -0,0 +1,120 @@
+  - StringUtils 클래스로 NEWLINE 상수 값을 옮긴다
+    - ```java
+          public static final String NEWLINE = System.getProperty("line.separator"); 
+         ```
````
```suggestion
    ```java
    public static final String NEWLINE = System.getProperty("line.separator"); 
    ```
```

이렇게 해도 잘 나올거에요!

jihye-woo replied:
여기서는 고쳐주신대로 - 이 없는 편이 깔끔한 것 같아요! 그리고 마크다운에서 ``` 는 indentation이 맞지 않을 때 가끔 이상하게 보이더라구요 마찬가지로 고쳐주신대로 처음과 끝 부분의 indentation을 잘 맞춰야겠습니다!
체크리스트까지 봐주시다니..ㅠ😳 감사합니다!!🙇‍♀️🙇‍♀️

ksundong commented on checklist/미션4.md:
```diff
@@ -0,0 +1,120 @@
+
+### 7. 최종 점검
+
+- [ ] main() 메소드를 실행
```
점검이 안되었나요? 👀

jihye-woo replied:
점검을 했는데 체크를 안했네요..ㅠㅠ ✔✔✔ 체크했습니다!

ksundong commented on src/main/java/net/woody/chess/Board.java:
```diff
@@ -15,47 +15,80 @@
+        initWhitePieces();
+    }
+
+    public void initBlackPieces() {
+        Rank blackPieces = getRank(Piece.BLACK_PIECES_RANK);
+
+        blackPieces.add(Piece.createBlackRook());
+        blackPieces.add(Piece.createBlackKnight());
+        blackPieces.add(Piece.createBlackBishop());
+        blackPieces.add(Piece.createBlackQueen());
+        blackPieces.add(Piece.createBlackKing());
+        blackPieces.add(Piece.createBlackBishop());
+        blackPieces.add(Piece.createBlackKnight());
+        blackPieces.add(Piece.createBlackRook());
+        size += 8;
+
+        Rank blackPawns = getRank(Piece.BLACK_PAWN_RANK);
+
+        for (int i = 0; i < BOARD_LENGTH; i++) {
+            blackPawns.add(Piece.createBlackPawn());
+        }
+        size += 8;
+    }
```
이 메서드 너무 길지 않나요?
size를 더해주는 것도 뭔가 이상하구요.
Pawn을 추가하는 것과 다른 기물들을 추가하는 부분을 나눠보는건 어때요?

ksundong replied:
하드코딩도 지양해주세요!

jihye-woo replied:
#### initialize() 분리
- 말씀하신대로 색상뿐만 아니라 기물 타입에 따라 initialize() 메소드를 조금 더 세분화시켜보았습니다
 #### size() 수정
 - size() 호출 시, 모든 Rank를 순회하며 모든 size 값을 더해 리턴하는 방식으로 수정하였습니다!

ksundong commented on src/main/java/net/woody/chess/Game.java:
```diff
@@ -1,23 +1,23 @@
-            }
-        } while (ongoing);
-        sc.close();
+        try (Scanner sc = new Scanner(System.in)) {
```
try with resources의 사용 좋습니다 👍

ksundong commented on src/main/java/net/woody/chess/Game.java:
```diff
@@ -1,23 +1,23 @@
-        sc.close();
+        try (Scanner sc = new Scanner(System.in)) {
+            do {
+                CommandUtils.printMenu();
```
오 `CommandUtils`라는게 있군요.

jihye-woo replied:
넵! 메인이 너무 길어져서 Command 관련 메소드들은 `CommandUtils`로 분리시켰습니다!😀

ksundong commented on src/main/java/net/woody/chess/Game.java:
```diff
@@ -1,23 +1,23 @@
+            do {
+                CommandUtils.printMenu();
+                String command = sc.nextLine();
+                ongoing = CommandUtils.isGameOver(command);
```
👍

ksundong commented on src/main/java/net/woody/chess/Game.java:
```diff
@@ -1,23 +1,23 @@
+                ongoing = CommandUtils.isGameOver(command);
+                if (ongoing) {
+                    board.initialize();
+                    System.out.println(board.print());
```
`print`를하는데 안에서 `print`를 하네요! 메서드 네이밍을 한 번 고려해봅시다.

jihye-woo replied:
코드를 다시 읽어보니까 확실히 이상하네요👀 디테일하게 봐주셔서 감사합니다!!🙇‍♀️🙇‍♀️ 
⁕ `showBoard()`로 수정했습니다!

ksundong commented on src/main/java/net/woody/chess/Representation.java:
```diff
@@ -0,0 +1,17 @@
+
+public enum Representation {
+
+    PAWN('p'), KNGINT('n'), ROOK('r'), BISHOP('b'), QUEEN('q'), KING('k');
```
```suggestion
    PAWN('p'),
    KNGINT('n'),
    ROOK('r'),
    BISHOP('b'),
    QUEEN('q'),
    KING('k');
```

Enum에 속성이 있을 때는 요렇게 써주세요~

jihye-woo replied:
Enum에 속성 정보가 char 하나라서 신경써주지 않았는데 들켰네요..!👀 수정했습니다!😁

ksundong commented on src/main/java/net/woody/pieces/Color.java:
```diff
@@ -7,10 +7,10 @@
     BLACK(piece -> Character.toUpperCase(piece));
 
-    private Function<Character, Character> represent;
+    private Function<Character, Character> representation;
```
오 함수를 사용해주셨네요!

jihye-woo replied:
네엡! Color 정보와 대소문자 변경 기능을 하나로 묶고 싶어서 함수로 구현하게 되었습니다!

ksundong commented on src/main/java/net/woody/pieces/Color.java:
```diff
@@ -25,8 +25,8 @@ public static Color value(String color) {
 
-    public Character representation(char pieceFirstLetter) {
-        return represent.apply(pieceFirstLetter);
+    public Character represent(char pieceFirstLetter) {
```
다른 네이밍은 어떨까요? 차라리 `getRepresentation`을 사용해보는 것도 좋을 것 같습니다.

ksundong commented on src/main/java/net/woody/utils/CommandUtils.java:
```diff
@@ -1,6 +1,9 @@
-public class CommandUtil {
+public class CommandUtils {
+
+    private CommandUtils() {
+    }
```
이 클래스의 기본생성자를 `private`으로 만들어주신 이유가 있나요?

jihye-woo replied:
utils 클래스들은 따로 객체로 생성해줄 필요가 없을 것 같아서 private으로 설정해주었습니다..!

ksundong commented on src/main/java/net/woody/utils/CommandUtils.java:
```diff
@@ -10,7 +13,7 @@ public static void printMenu() {
     }
 
-    public static boolean readCommand(String command) {
+    public static boolean isGameOver(String command) {
```
뭔가 하는 역할은 game이 끝났을 때 호출할 것 같은데
좀 더 다른 네이밍을 고민해봐도 좋을 것 같아요!

ksundong commented on src/main/java/net/woody/utils/StringUtils.java:
```diff
@@ -0,0 +1,17 @@
+    }
+
+    public static String appendNewLine(String line) {
+        return new StringBuilder()
+                .append(line)
+                .append(NEWLINE)
+                .toString();
```
```suggestion
        return new StringBuilder(line)
                .append(NEWLINE)
                .toString();
```

이렇게도 된답니다!

jihye-woo replied:
오오 감사합니다!! 저렇게 수정해야겠어요! 코드 한 줄 줄어들었을 뿐이지만 정말 기분이 좋네용✨✨😊

ksundong commented on src/test/java/net/woody/chess/BoardTest.java:
```diff
@@ -13,70 +13,68 @@
+    @Test
+    @DisplayName("보드를 초기화한 후, 체스말이 총 32개여야 한다.")
+    void checkTheNumberOfPieces() {
+        assertThat(board.size()).isEqualTo(32);
```
여기도 32라는 숫자 대신에 변수로 추출해줘도 잘 읽힐 것 같아요!
이 부분은 수정 안하셔도 됩니다!

jihye-woo replied:
앗 그렇군요! 확실히 32이라는 숫자가 무엇을 뜻하는지 변수명으로 좀 더 확실하게 표현해주는 것도 좋을 것 같네요✨지금은 참고만 하고 넘어가겠습니다!!👀 짚어주셔서 감사해요🙇‍♀️🙇‍♀️

ksundong commented on src/test/java/net/woody/chess/BoardTest.java:
```diff
@@ -13,70 +13,68 @@
-
-        Pawn secondPawn = new Pawn();
-        addThePawnProperly(secondPawn, 1);
+        addThePieceProperly(Piece.createWhitePawn(), 2, 0);
+        addThePieceProperly(Piece.createWhiteKing(), 2, 1);
```
뒤의 숫자들만 봐서는 무슨 메서드인지 유추하기가 어렵네요.
왜 2고, 0이고 1인지 잘 모르겠어요.

jihye-woo replied:
- 테스트 메소드 `findPawnOnTheBoard()`의 로직이 테스트 목적과 살짝 맞지 않아서 수정하는 과정에서 헬퍼 메소드였던 `addThePieceProperly()`가 더 이상 필요하지 않게 되어서 제거해주었습니다!

- 만약 해당 메소드를 삭제하지 않았다면, 메소드명에 '특정위치에 넣어준다'는 의미를 추가해서 메소드가 인자들로 어떤 일을 할건지를 명확히 해주는 쪽으로 수정했을 것 같아요..! (그렇게 하면 디온이 말했던 문제가 해결이 될 수 있을까요?)

ksundong replied:
인자를 변수로 추출하는 편이 더 읽기 좋을 것 같습니다.

jihye-woo replied:
그렇군요! 다음 단계에서 반영하도록 하겠습니다 감사합니다!

ksundong commented on src/test/java/net/woody/chess/BoardTest.java:
```diff
@@ -13,70 +13,68 @@
+    @DisplayName("체스말이 존재하지 않는 위치에서, 폰을 찾으려고 시도하면 에러가 발생해야 한다.")
     void findPawnNotOnTheBoard() {
-        assertThatThrownBy(() -> board.findPawn(0, 0))
+        assertThatThrownBy(() -> board.findPiece(4, 0))
                 .isInstanceOf(ArrayIndexOutOfBoundsException.class)
                 .hasMessageContaining("0 is out of range!");
```
코드가 바뀌었는데 테스트가 통과했나요?

jihye-woo replied:
`@BeforeEach`에 보드 초기화하는 코드를 추가하게 되면서 **(0, 0) 자리에는 체스말이 존재하는 상태**가 되었습니다..!
해당 메소드는 체스말이 존재하지 않는 위치에서 체스말을 찾으려고 시도하는 메소드이기 때문에 **초기화 직후에도 체스말이 존재하지 않는 (4, 0)로 변경**해주었습니다..!
테스트는 통과되었습니다!

jihye-woo replied:
여기도 메소드 네이밍을 수정해줄 필요가 있겠군요! `findPawnNotOnTheBoard`에서 `findPieceNotOnTheBoard`로 수정해주었습니다!

ksundong replied:
"0 is out of range!"
이 문구는 뒤의 0 때문에 발생하는 것인가요?
살짝 헷갈리는 부분이 있네요~

jihye-woo replied:
훔 그렇군요..! 차라리 out of range! 라는 메세지가 더 적절할까요?

ksundong replied:
특정해주는 것이 불필요할 수도 있을 것 같아요~

ksundong commented on src/test/java/net/woody/chess/BoardTest.java:
```diff
@@ -13,70 +13,68 @@
-        board.add(newPawn);
-        assertThat(sizeBeforeAddThePawn + 1).isEqualTo(board.size());
-        assertThat(newPawn).isEqualTo(board.findPawn(rank, newPawnIdx));
+    private void addThePieceProperly(Piece newPiece, int rank, int file) {
```
The가 굳이 안붙어도 될 것 같아요.

jihye-woo replied:
넵넵 네이밍 수정하도록 하겠습니다! 
정말 필요한 정보(키워드)들만 메소드에 담는게 좋을 것 같네요🤔

ksundong commented on src/test/java/net/woody/chess/BoardTest.java:
```diff
@@ -13,70 +13,68 @@
+    private void addThePieceProperly(Piece newPiece, int rank, int file) {
+        int sizeBeforeAddThePiece = board.size();
+        board.add(newPiece, rank);
+        assertThat(sizeBeforeAddThePiece + 1).isEqualTo(board.size());
+        assertThat(newPiece).isEqualTo(board.findPiece(rank, file));
```
actual, expected의 순서로 작성해주세요!

jihye-woo replied:
ㅠㅠ 넹넹! 바로 수정해야겠네요! 감사합니다🙇‍♀️🙇‍♀️

ksundong commented on src/test/java/net/woody/pieces/PieceTest.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @Test
+    @DisplayName("흰색으로 생성된 체스말은 흰색이어야 하고, 검은색으로 생성했을 때는 검은색이어야 한다.")
```
Pawn만 검사하는데 설명이 조금 다르네요!

jihye-woo replied:
그렇네요! 엄밀히 말하면 현재 테스트 메소드는 Pawn만 검사하고 있는 상태니까 DisplayName을 조금 더 정확하게 고치는게 좋겠습니다!😊😊

ksundong reviewed (changes requested):
안녕하세요~ Woody 리뷰어 Dion입니다. 4단계 미션까지 작성하시느라 고생하셨어요! 👍
몇가지 수정사항과, 코멘트를 남겼습니다. 확인 후에 반영해주세요.
전반적으로 깔끔한 코드여서 말씀 드린 부분만 수정해주시면 바로 다음단계 진행하셔도 될 것 같아요.

## 리뷰 사항

- 메서드의 역할 분리를 고민해주세요.
- 메서드 네이밍을 좀 더 고민해주세요.
- enum에 속성이 있을 땐 개행을 해주세요.
- 테스트 코드도 읽기 좋게 수정해주세요.
- 테스트 코드가 변경되었는데 통과하는 코드가 있습니다. 이를 개선해주세요.
- 메서드명이나 변수명에 관사는 붙이지 않는게 좋을 것 같아요.
- assertJ 테스트 메서드는 actual, expected 순서로 작성해주세요.

고생하셨습니다! 궁금하시거나, 이해가 잘 안되거나, 도움이 필요한 부분은 코멘트 남겨주세요.
커밋메시지를 잘 작성해주시고, PR도 잘 작성해주셨습니다. 👍

jihye-woo committed:
refactor : 테스트 클래스 DisplayName 수정

  > DisplayName 수정
    - 메소드 로직을 제대로 설명하지 못하는 부분 수정

jihye-woo committed:
edit : BoardTest 테스트 메소드 수정

  > findPawnOnTheBoard() 테스트 클래스
    - 테스트 의도와 맞지 않은 로직 수정
    - 특정 위치의 체스말을 가져올 수 있는지 테스트

  > initializeBoard() 테스트 클래스 삭제
    - printBoard()에서도 이중으로 Pawn 라인에 대한 테스트를 진행해주므로 필요성이 없어짐

  > addPieceProperly() 헬퍼 메소드 삭제
    - 더 이상 사용하지 않는 메소드이므로 삭제

jihye-woo committed:
refactor : StringUtils 클래스 메소드 리펙토링

  > appendNewLine() 메소드 더 간결하게 수정

jihye-woo committed:
refactor : Representation 클래스 오타 및 코드 스타일 수정

  > KNGINT에서 KNIGHT로 오타 수정

  > Enum 상수에 속성 존재할 경우, 세로로 열거

jihye-woo committed:
docs: 미션 5 체크리스트 업데이트

  > main() 메소드 실행 완료 후 체크

  > 마크다운 내 java 코드 작성 부분 수정

jihye-woo committed:
refactor : 메소드 명 수정

  > Color 클래스
    - represent() → getRepresentation()

  > Board 클래스
    - print() → showBoard()

  > CommandUtils 클래스
    - isGameOver() → isGameContinued()
    - 리턴값을 고려했을 때 isGameContinued가 더 적절

jihye-woo committed:
refactor : Board 클래스 size() 수정, 불필요한 메소드 제거

  > 하드코딩 제거
    - size() 호출 시, 모든 Rank를 순회하며 모든 size 값을 더해 리턴

  > 불필요한 메소드 제거
    - getWhitePawnsResult()와 getBlackPawnsResult() 삭제

jihye-woo committed:
refactor : Board 클래스 메소드 수정 및 삭제

  > 하드코딩 제거
    - size() 호출 시, 모든 Rank를 순회하며 모든 size 값을 더해 리턴

  > initialize() 메소드 수정
    - Pawn과 Pawn이 아닌 체스말 초기화 로직 분리
    - 헬퍼 메소드 접근 지시자 private 설정

  > 불필요한 메소드 제거
    - getWhitePawnsResult()와 getBlackPawnsResult() 삭제

jihye-woo committed:
edit : Piece 클래스 equals() hashCode() 재정의

  > Piece 클래스는 Value Object로 간주
    - 같은 값을 가진 Value Object에 대해서 식별성 제거해야 함
    - 동일성(참조값)이 아닌 논리적 동치성을 비교해야 함
    - equals() : 값을 비교할 수 있도록 재정의
    - hashCode() : 같은 값을 가진 객체가 같은 hash 값을 가질 수 있도록 재정의

  > findPawnOnTheBoard()에서 isEqualTo() 적용
    - 값이 같은 Piece 객체에 대해 equals() 결과로 true를 기대할 수 있게 됨

jihye-woo commented:
@ksundong 
디온이 코멘트 주신부분들 수정해보았습니다! 
커밋 메세지만으로도 수정 사항을 이해하실 수 있도록 구체적으로 작성하긴 했지만 그래도! 한눈에 보실 수 있도록 수정 내용을 간단히 정리해보았습니다😀 도움이 될까 싶어서 작성해놓은 글이라 다 읽으실 필요는 없습니다! 참고만 해주셔도 좋습니다👍

#### 🙋‍♀️ 질문
 1. 현재 팩토리 메소드가 체스말의 색깔별로 타입별로 존재하다보니 10개가 넘어가게 되었는데, 원래도 이렇게 정적 팩토리 메소드를 하나의 클래스 안에 10개 이상씩 두는 경우가 많나요? 익숙하지 않은 방식이라 그런지 어색하네요..!ㅠㅠ 클래스가 길어지고 비슷한 기능들을 가진 메소드가 열거되어있다는 느낌이 듭니다..! 
 
#### 😫 마지막 두 refactor 커밋 메세지 내용이 중복되었습니다ㅠㅠ  하나의 커밋으로 생각해주시면 감사드리겠습니다..!


---

### 테스트 관련 수정
#### 1. `@DisplayName` 수정
  - Test 메소드의 `@DisplayName` 한 번씩 더 확인하고 수정했습니다.
#### 2. BoardTest 클래스의 findPawnOnTheBoard() 테스트 메소드 로직 수정
  - 테스트 메소드의 의도와 메소드 로직이 일치하지 않아 메소드 로직을 수정했습니다.
### 로직 수정 및 추가 구현
#### 1. size() 로직 수정
  - Board 클래스에서 하드코딩으로 size 값을 업데이트하는 부분을 삭제하였습니다.
  - size()를 호출할 때, 모든 Rank를 순회하면서 size 값을 더해 리턴하는 방식으로 로직을 수정하였습니다.
#### 2. Piece 클래스 equals() 재정의
  - 값 객체로서 논리적 동치성을 비교해주기 위해서 equals()를 재정의해주었습니다. (hashCode도 함께 재정의)
### 리펙토링
#### 1. StringUtils 리펙토링
   - 코드 한 줄 줄일 수 있는 방법을 알려주셔서 적용했습니다
#### 2. Representation enum 클래스 수정
  - Enum 상수를 세로로 열거해놓았습니다!
#### 3. 미션 4 체크리스트 업데이트
  - 달성한 요소들 모두 체크 표시하고, 마크다운 내부에서 java 코드 부분도 말씀하신대로 수정했습니다
#### 4. 메소드 명 수정
  - 조금 더 (제 기준에서😳) 메소드 이름으로 메소드 로직을 명확하게 설명할 수 있도록 수정했습니다.
#### 5 . 불필요한 메소드 제거
  -  지금 단계에서 필요하지 않은 메소드들은 제거해주었습니다.

---

읽어주셔서 감사합니다😀

ksundong commented on src/main/java/net/woody/chess/Board.java:
```diff
@@ -35,37 +30,71 @@ public Rank getRank(int rank) {
 
     public int size() {
-        return size;
+        int numOfPieces = 0;
+        for (Rank rank : board) {
+            numOfPieces += rank.size();
+        }
+        return numOfPieces;
```
매번 size를 확인할 때마다 전체를 탐색할 것 같은데, 이 부분은 좀 더 고민해봐야 할 것 같아요.

jihye-woo replied:
넵 일단 하드코딩은 제거했지만 size()를 호출할 때마다 연산을 해야해서 조금 더 고민해봐야할 것 같습니다..!ㅜㅠ

ksundong commented on src/main/java/net/woody/chess/Board.java:
```diff
@@ -35,37 +30,71 @@ public Rank getRank(int rank) {
 
-    public String print() {
+    public String showBoard() {
         StringBuilder sb = new StringBuilder();
```
sb보다 좋은 네이밍을 한 번 고민해봅시다.

ksundong commented on src/main/java/net/woody/pieces/Piece.java:
```diff
@@ -0,0 +1,97 @@
+        return new Piece(Color.WHITE, Representation.KING.value());
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Piece piece = (Piece) o;
+        return representation == piece.representation && color == piece.color;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(color, representation);
+    }
```
`equals`, `hashCode` 오버라이딩 해주신 이유가 있을까요?

jihye-woo replied:
Piece를 value object로 취급하게되면 값으로서 비교를 해주어야하는데, equals를 재정의하지 않으면 같은 값을 가진 Piece 객체를 다르게 식별할 수 있겠더라구요..!

(+) 리뷰어이신 소니의 글이 많은 참고가 되었습니다👀
  - [[Java] equals() & hascode() 메서드는 언제 재정의해야 할까?](https://velog.io/@sonypark/Java-equals-hascode-%EB%A9%94%EC%84%9C%EB%93%9C%EB%8A%94-%EC%96%B8%EC%A0%9C-%EC%9E%AC%EC%A0%95%EC%9D%98%ED%95%B4%EC%95%BC-%ED%95%A0%EA%B9%8C)

ksundong replied:
저는 이정도의 미래 지향적인 코드는 상관 없다고 봅니다.
다만 동일성, 동등성에 대해서 꼭 고민해보시기 바랍니다!

ksundong commented on src/main/java/net/woody/utils/CommandUtils.java:
```diff
@@ -10,7 +13,7 @@ public static void printMenu() {
     }
 
-    public static boolean readCommand(String command) {
+    public static boolean isGameContinued(String command) {
```
저라면 그냥 isStart를 쓸 것 같아요.

jihye-woo replied:
그렇군요! `isStart`가 간단하고 좋네요👍

ksundong commented on src/test/java/net/woody/chess/BoardTest.java:
```diff
@@ -13,70 +14,59 @@
+    void findPawnOnTheBoard() {
+        assertAll(
+                () -> {
+                    Piece shouldBeBlackRook = board.findPiece(0, 0);
```
shouldBe 같은 이름은 제외해줍시다~

jihye-woo replied:
테스트 클래스의 변수에 한해서는 가능할 수도 있다고 생각했는데 역시 지양하는게 좋겠군요! 디테일한 코멘트 감사합니다!😊

ksundong commented on src/test/java/net/woody/chess/BoardTest.java:
```diff
@@ -13,70 +14,59 @@
+                        appendNewLine("rnbqkbnr");
 
-        String actualResult = board.print();
+        String actualResult = board.showBoard();
         assertThat(expectedResult).isEqualTo(actualResult);
```
expected, actual로 변경해줘도 될 것 같아요.

jihye-woo replied:
그렇군요! Result라는 표현이 굳이 들어가지 않아도 expected, actual이라는 네이밍 만으로 충분히 코드 흐름이 전달될 것 같네요 감사합니다!😊

ksundong reviewed (approved):
Woody! 수정해주시느라 고생하셨습니다.
저는 정적 팩토리 메서드를 타입별로 생성해주고, 색을 인자로 받는 방식으로 만들 것 같습니다.
제가 개발할 때에는 정적 팩토리 메서드를 그렇게 많이 만들어주진 않는 것 같아요.
하지만 분명히 장점이 있는 방식이니만큼 익숙해지시면 좋을 것 같습니다.

바로 다음 단계 진행해주시면 될 것 같구, 좀 더 생각해 볼 부분에 대해 코멘트 남겼으니 확인해주세요!
고생하셨습니다~
